core/commands/ptp.go

```diff
@@ -0,0 +1,392 @@
+package commands
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"text/tabwriter"
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+	core "github.com/ipfs/go-ipfs/core"
+
+	ma "gx/ipfs/QmcyqRMCAXVtYPS4DiBrA7sezL9rRGfW8Ctx7cywL4TXJj/go-multiaddr"
+)
+
+// PTPListenerInfoOutput is output type of ls command
+type PTPListenerInfoOutput struct {
+	Protocol string
+	Address  string
+}
+
+// PTPStreamInfoOutput is output type of streams command
+type PTPStreamInfoOutput struct {
+	HandlerID     string
+	Protocol      string
+	LocalPeer     string
+	LocalAddress  string
+	RemotePeer    string
+	RemoteAddress string
+}
+
+// PTPLsOutput is output type of ls command
+type PTPLsOutput struct {
+	Listeners []PTPListenerInfoOutput
+}
+
+// PTPStreamsOutput is output type of streams command
+type PTPStreamsOutput struct {
+	Streams []PTPStreamInfoOutput
+}
+
+// PTPCmd is the 'ipfs ptp' command
+var PTPCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Libp2p stream mounting.",
+		ShortDescription: `
+Create and use tunnels to remote peers over libp2p
+
+Note: this command is experimental and subject to change as usecases and APIs are refined`,
+	},
+
+	Subcommands: map[string]*cmds.Command{
+		"ls":      ptpLsCmd,
+		"streams": ptpStreamsCmd,
+		"dial":    ptpDialCmd,
+		"listen":  ptpListenCmd,
+		"close":   ptpCloseCmd,
+	},
+}
+
+var ptpLsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "List active p2p listeners.",
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("headers", "v", "Print table headers (HandlerID, Protocol, Local, Remote).").Default(false),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		err = checkEnabled(n)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if !n.OnlineMode() {
+			res.SetError(errNotOnline, cmds.ErrClient)
+			return
+		}
```

Comment:
IMO I would compress code above into once function (returning node or error in case those checks fail) as it is used everywhere.